intel_extension_for_transformers/backends/neural_engine/graph/application/ChatGLM/chatglm.cpp

```diff
@@ -0,0 +1,1250 @@
+                               v_cache->nb[2], 0); // [heads, head_size, klen]
+    
+    ne_tensor *attn_scores = ne_mul_mat(gctx, key_layer, query_layer); // [heads, qlen, klen]
+    
+    if (n_past == 0) {
+        // build attention mask for context input
+        ne_tensor *inf = ne_new_tensor_3d(gctx, attn_scores->type, 1, qlen - 1, num_attention_heads, NE_SIZE_CALC);
+        ne_set_f32(inf, -INFINITY);
+
+        ne_tensor *masked_attn_scores =
+            ne_view_3d(gctx, attn_scores, 1, qlen - 1, num_attention_heads, qlen * ne_element_size(attn_scores),
+                         qlen * qlen * ne_element_size(attn_scores), (qlen - 1) * ne_element_size(attn_scores));
+        
+        ne_build_forward_expand(&ctx->gf, ne_cpy(gctx, inf, masked_attn_scores));
+    }
+    attn_scores = ne_scale_inplace(gctx, attn_scores, ne_new_f32(gctx, 1.f / std::sqrt(head_size)));
+    
+    ne_tensor *attn_probs = ne_soft_max_inplace(gctx, attn_scores); // [heads, qlen, klen]
+    
+
+    ne_tensor *context_layer = ne_mul_mat(gctx, value_layer, attn_probs); // [heads, qlen, head_size]
+    
+    context_layer = ne_cont(gctx, ne_permute(gctx, context_layer, 0, 2, 1, 3));
+    
+    context_layer = ne_reshape_2d(gctx, context_layer, hidden_size, qlen);
+    
+
+    ne_tensor *attn_output = dense.forward(ctx, context_layer);
+    return attn_output;
+}
+
+ne_tensor *GLMBlock::forward(ModelContext *ctx, ne_tensor *hidden_states, int n_past, int n_ctx) const {
+    ne_context *gctx = ctx->ctx_b.get();
+
+    ne_tensor *alpha = ne_new_f32(gctx, std::sqrt(2.f * num_hidden_layers));
+
+    ne_tensor *attn_input = input_layernorm.forward(ctx, hidden_states);
+    ne_tensor *attn_output = attention.forward(ctx, attn_input, n_past, n_ctx);
+    ne_build_forward_expand(&ctx->gf, attn_output);
+    attn_input = ne_scale_inplace(gctx, attn_input, alpha);
+    
+    hidden_states = ne_add_inplace(gctx, attn_input, attn_output);
+    
+    ne_tensor *mlp_input = post_attention_layernorm.forward(ctx, hidden_states);
+    ne_tensor *mlp_output = mlp.forward(ctx, mlp_input);
+    ne_build_forward_expand(&ctx->gf, mlp_output);
+    mlp_input = ne_scale_inplace(gctx, mlp_input, alpha);
+    
+    ne_tensor *output = ne_add_inplace(gctx, mlp_input, mlp_output);
+    
+
+    return output;
+}
+
+ChatGLMModel::ChatGLMModel(ModelContext *ctx, const ChatGLMConfig &config)
+    : word_embeddings(ctx, config.vocab_size, config.hidden_size), final_layernorm(ctx, config.hidden_size) {
+    layers.reserve(config.num_hidden_layers);
+    for (int layer_id = 0; layer_id < config.num_hidden_layers; layer_id++) {
+        layers.emplace_back(ctx, config.hidden_size, config.num_attention_heads, config.num_hidden_layers,
+                            config.max_length);
+    }
+}
+
+ne_tensor *ChatGLMModel::forward(ModelContext *ctx, ne_tensor *input_ids, int n_past, int n_ctx) const {
+    ne_context *gctx = ctx->ctx_b.get();
```

Comment:
model structure should be under models/

Reply:
The https://github.com/intel/intel-extension-for-transformers/pull/107 PR will refactor the model.